source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1157,7 +1157,12 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::onHostHealthFailure(
     if (container != config.host_tcp_conn_pool_map_.end()) {
       for (const auto& pair : container->second.pools_) {
         const Tcp::ConnectionPool::InstancePtr& pool = pair.second;
-        pool->drainConnections();
+        if (host->cluster().features() &
+            ClusterInfo::Features::CLOSE_CONNECTIONS_ON_HOST_HEALTH_FAILURE) {
```

Comment:
The code on L1170 closes the old style non-connection pool TCP connections: it was the refactoring to move tcp_proxy away from those to the TCP connection pool that broke this feature. As far as I can tell they're only used by the TCP stats sync now.
It would probably be best to convert the TCP stats sync to using a TCP connection pool, and delete all the code that supports the non-pooled connections. I looked at doing that, but it was a bit beyond what I think I'm able to do.

Reply:
@JonathanO Can you add a comment/TODO with that information, so the next person to see this doesn't have to re-research it again?

Reply:
Sure, I've added a comment about why it's there, and a TODO to suggest removing it.